test/functional/lib/ipfs-request-protocol-handlers.test.js

```diff
@@ -61,173 +61,339 @@ describe('modifyRequest.onBeforeRequest:', function () {
         // without web+ prefix (Firefox > 59: https://github.com/ipfs-shipyard/ipfs-companion/issues/164#issuecomment-356301174)
         it('should not be normalized if ipfs:/{CID}', async function () {
           const request = url2request('https://dweb.link/ipfs/?uri=ipfs%3A%2FQmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR%3FargTest%23hashTest')
-          expect(await modifyRequest.onBeforeRequest(request)).to.equal(undefined)
+
+          ensureRequestUntouched(await modifyRequest.onBeforeRequest(request))
         })
         it('should be normalized if ipfs://{CID}', async function () {
           const request = url2request('https://dweb.link/ipfs/?uri=ipfs%3A%2F%2FQmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR%3FargTest%23hashTest')
-          expect((await modifyRequest.onBeforeRequest(request)).redirectUrl).to.equal('https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?argTest#hashTest')
+
+          ensureCallRedirected({
+            modifiedRequestCallResp: await modifyRequest.onBeforeRequest(request),
+            MV2Expectation: 'https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?argTest#hashTest',
+            MV3Expectation: {
+              origin: '^https?\\:\\/\\/dweb\\.link\\/ipfs\\/\\?uri\\=ipfs%3A%2F%2FQmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR%3FargTest%23',
+              destination: 'https://ipfs.io/ipfs/QmbWqxBEKC3P8tqsKc98xmWNzrzDtRLMiMPL8wBuTGsMnR?argTest#'
+            }
+          })
         })
         it('should not be normalized if ipns:/{fqdn}', async function () {
           const request = url2request('https://dweb.link/ipfs/?uri=ipns%3A%2Fipfs.io%3FargTest%23hashTest')
-          expect(await modifyRequest.onBeforeRequest(request)).to.equal(undefined)
+
+          ensureRequestUntouched(await modifyRequest.onBeforeRequest(request))
         })
         it('should be normalized if ipns://{fqdn}', async function () {
           const request = url2request('https://dweb.link/ipfs/?uri=ipns%3A%2F%2Fipfs.io%3FargTest%23hashTest')
-          expect((await modifyRequest.onBeforeRequest(request)).redirectUrl).to.equal('https://ipfs.io/ipns/ipfs.io?argTest#hashTest')
+
+          ensureCallRedirected({
+            modifiedRequestCallResp: await modifyRequest.onBeforeRequest(request),
+            MV2Expectation: 'https://ipfs.io/ipns/ipfs.io?argTest#hashTest',
+            MV3Expectation: {
+              origin: '^https?\\:\\/\\/dweb\\.link\\/ipfs\\/\\?uri\\=ipns%3A%2F%2Fipfs\\.io%3FargTest%23',
+              destination: 'https://ipfs.io/ipns/ipfs.io?argTest#'
```

Comment:
:warning:  Not blocking this PR, but see https://github.com/ipfs/ipfs-companion/issues/1254 for follow-up work.